Add missing SIGRTMIN and SIGRTMAX constants and handling these signals to pcntl

DennisBirkholz opened the pull request:
Hi PHP folks,

please consider my patch that makes the "real time signals" specified by Posix available/usable in PHP.
Real time signals are just signals without a specific meaning and free to use by the user. Currently it is not possible to register a signal handler for them or use them safely as the `SIGRTMIN` and `SIGRTMAX` constants are not set.

I did not protect the changes by `#ifdef` guards as the `SIGRTMAX` constant was already used unprotected but I can easily add them.

Thx,
Dennis

laruence commented:
SIGRTMAX is checked by pcntl, but SIGRTMIN is not. and my MAC OS doesn't have them (checked by kill -l)

dennisbirkholz commented:
I added an ifdef guard to the SIGRTMIN constant, maybe you can merge it now (it's also rebased against current master).

cpriest commented:
Any chance this is going to get merged?

krakjoe commented:
Well, I would like too, but I'm thrown a bit by @laruence comment that they are not defined in OSX.

@laruence can you double check that please ?

If indeed neither are defined, then the patch will still break OSX, further, it would mean that OSX does not have a posix compliant implementation of signals ... which I find very strange, and makes me reluctant to do anything.

krakjoe commented:
Having waited more than a month for feedback, and having received nothing, I'm closing this PR as it would appear abandoned.

DennisBirkholz commented:
This PR is not abandoned, please reopen! I just can not provide any information about OSX behavior, but maybe some Apple-user can? Hopefully @laruence?

krakjoe commented:
Neither of them would appear to be defined on 10.12.3 (16D32), so both need to be ifdef'd for definition. In addition, the patch will still break on operating systems where they are not defined (as I mentioned a month ago) because they unconditionally reference them [here](https://github.com/php/php-src/pull/1457/files#diff-7185d47849fdb94217f22977d69e85b1R963) and again a little further down.

If you'd like to fix the patch so that it won't break stuff, I'll happily re-open it ... it would seem a month is long enough for you to have done that already.

DennisBirkholz commented:
Hi Joe,

sorry I did not get that you requested an action from me, sorry. But I don't get why SIGRTMAX can not be used like that. It gets defined in [ext/pcntl/php_signal.h](https://github.com/php/php-src/blob/master/ext/pcntl/php_signal.h#L28) on line 28 so it is safe to use in [pcntl.c](https://github.com/php/php-src/blob/master/ext/pcntl/pcntl.c). Or am I wrong?

Thanks,
Dennis

nikic commented:
@DennisBirkholz Thanks for pointing that out. The define looks very dubious to me -- if the system does not define SIGRTMAX by itself, it most likely will not have 64 signals.  The defines in that header should probably look something like this:

```
#ifndef NSIG
# ifdef SIGRTMAX
#  define NSIG (SIGRTMAX+1)
# else
#  define NSIG 32
# endif
#endif
```

(Though I'm not totally sure if NSIG, if it is defined, always includes RT signals. On my system this is the case, but I'm not sure how general that is.)

In any case, the PHP-exposed constants SIGRTMIN and SIGRTMAX should either both be defined or both not be defined. Having SIGRTMAX always defined while SIGRTMIN is not is weird.

Furthermore, it should be noted that SIGRTMIN and SIGRTMAX are not actually constants. It is spec'd that they may change at runtime. As such, it may be problematic to expose them to PHP as constants. (I'm not familiar with whether the values changes in practice or not.)

nikic commented on ext/pcntl/pcntl.c:
```diff
@@ -965,7 +969,7 @@ PHP_FUNCTION(pcntl_signal)
 		 * pre-allocate a few records for recording signals */
 		int i;
-		for (i = 0; i < 32; i++) {
+		for (i = 0; i < MAX(NSIG-1, SIGRTMAX); i++) {
```
This condition should be `i < MAX(NSIG, SIGRTMAX+1)`, otherwise SIGRTMAX cannot be set.

DennisBirkholz commented:
@nikic I changed the condition-error and rebased onto current master.

> Furthermore, it should be noted that SIGRTMIN and SIGRTMAX are not actually constants. It is spec'd that they may change at runtime. As such, it may be problematic to expose them to PHP as constants. (I'm not familiar with whether the values changes in practice or not.)

As I read it from the signals man page they only change if the threading library of the glibc is changed. So that should happen really seldom on the same machine.

nikic commented:
Merged as https://github.com/php/php-src/commit/d0c8aba3d868d59a75489addb056d6be5b0b5f47, with some adjustments to NSIG handling (as discussed above) and conditional definition for SIGRTMAX.

Thanks!

DennisBirkholz commented:
Thanks for merging, @nikic!
